working/macros/feature-specification.md

```diff
@@ -45,6 +45,14 @@ Allowing deep introspection like this in cases where a macro needs it while
 ensuring that users can understand the system and tools can implement it
 efficiently is a central challenge of this proposal.
 
+#### Default Constructors
+
+Default constructors are not introduced until after phase 2 (the phase which
+might introduce generative constructors). They should not appear in
+introspection results until phase 3. If, after phase 2 a class still has no
+generative constructor, then the default one should be added, and it should
+be visible for introspection in phase 3.
```

Comment:
I wish we had a clearer principle for what introspection yields. Is it what the user wrote in the source code, or is it what abstract semantic model that desugars to? It seems like we keep stumbling into corners of the language that raise that question:

1. Does introspection show the implicit getter/setter introduced by a variable declaration?
2. Do primary constructors show up as just another constructor or not?
3. Do default constructors show up in introspection?

I suspect we will run into this problem more in the future as we add more declaration-level syntactic sugar features like (potentially) data classes, value types, etc.

The principle I lean towards is that introspection should show what the code is written to say and not what it happens to mean after desugaring. That would imply that default constructors don't show up in introspection at all even in phase 3.

I do realize that default constructors are different from the other desugarings because it's conditional: you only get one if no other constructors are defined. Maybe that warrants special treatment where default constructors never show up in introspection even if we decide that synthetic getters, synthetic setters, and primary constructors do.

Reply:
> The principle I lean towards is that introspection should show what the code is written to say and not what it happens to mean after desugaring.

I don't think this is what macro authors want. They want to know how a class can be _used_, not how it was written, generally.

> I wish we had a clearer principle for what introspection yields.

I think it should yield the most complete view of the thing being introspected over possible. That is what macro authors want, they don't really care what introduced the members. From a user perspective, it also shouldn't be breaking to move a constructor to a primary constructor, or visa-versa, and if macros can only see one or the other, or have to explicitly handle primary constructors in a special way, I think it makes it less likely that they can successfully handle them interchangeably.

This probably does mean showing implicit getters/setters, although we would need to highlight whether they were derived from a field or not.

Default constructors are I think unique, because they only are added if no generative constructor is added, and that doesn't play well with the phases.